### ISSUE TITLE
`model.encode` hangs after the last batch computation

### ISSUE THREAD

**Issue by nleroy917** [UKPLab/sentence-transformers#1677]
Hello,

I've been using `sentence-transformers` for a bit, and everything is great. I've noticed, however, that when I am batch-computing embeddings for a large corpus of sentences, the computation seems to hang after the last batch. I see a slightly related issue in #465, but I don't think this is that. I'm using `ipynb` in VSCode.

The progress bar claims that the encodings are complete after about 15 minutes:
<img width="926" alt="image" src="https://user-images.githubusercontent.com/41063083/186964811-fe0bbc9d-5727-4707-831c-ffea18e87951.png">

However, the cell is still running long after the encodings are "done". I am wondering what could be happening here:
<img width="935" alt="image" src="https://user-images.githubusercontent.com/41063083/186965694-14aa0833-d5d2-4732-bf8b-171c67179e01.png">

Is there any way to watch the progress of the post-processing? Here is the code I am executing. I'm encoding ~1 million sentences and the maximum sentence length is around 400 characters. I'm running on macOS with an M1 max chip. I know `pytorch` is just getting around to supporting M1 architectures.

```python
descriptions = [row['ItemDescription'] for row in rows_reduced]
description_embeddings = model.encode(
    descriptions,
    show_progress_bar=True, 
    convert_to_tensor=True,
    device="mps" # This is for Apple M1 chips. change to "cpu" or "cuda" for anything else (i.e. PC)
)
```

**Comment by vecorro** [comment]
Hi @nleroy917 I'm having the same issue with a list of 7.5M short (less that 50 words) sentences. I'm using a V100 GPU and after the bar indicates 100% completion, both CPU and GPU go to nearly 0% usage. The cell never finishes executing (at least not in the next 7-8 hours after the 100% completion) Did you find a fix for this problem?

**Comment by Hveemos** [comment]
Hi! I can confirm that I run into the same problem using: 
Apple M1 Max
64 GB
Ventura 13.1

Model: sentence-transformers/msmarco-distilbert-base-tas-b

transformers              4.24.0
pytorch                   1.13.0
sentence-transformers     2.2.2

When I try on a smaller sample (10000) it actually completes well. But when I increase the sample to 100000 it stops at 100% but never really finish.

**Comment by mebrunet** [comment]
Same problem

Model: sentence-transformers/all-MiniLM-L6-v2

Apple M1 Max, 32 GB, macOS 13.3.1

python=3.10.10=h3ba56d0_0_cpython
pytorch=2.0.0=py3.10_0
sentence-transformers=2.2.2=pyhd8ed1ab_0
transformers=4.27.4=py_0

**Comment by MaxBelitsky** [comment]
@nleroy917 I had the same issue recently and managed to solve this by removing `convert_to_tensor=True` and setting `convert_to_numpy=True`. I hope it helps.

**Comment by nleroy917** [comment]
@MaxBelitsky thanks. I'll try this out.

Does this affect downstream code, as you are not receiving the same type of object anymore? I've yet to really push the `encode` function with large batch sizes since opening this issue. But, I'm actually converting my vectors to lists later down:
```python
vector=p["embedding"].tolist(),
```
before stuffing into Qdrant. Wondering if a `np.array().tolist()` call is faster than `torch.Tensor().tolist()`

**Comment by MaxBelitsky** [comment]
@nleroy917 it does as it returns numpy array instead of torch tensor. If you need your embeddings in the tensor format, you can convert them after encoding. Let me know if this workaround helped.

**Comment by mebrunet** [comment]
Yeah it seems like it's the call to `torch.stack` that is very slow / hanging [see here](https://github.com/UKPLab/sentence-transformers/blob/d928410803bb90f555926d145ee7ad3bd1373a83/sentence_transformers/SentenceTransformer.py#L194-L195). 
If you set both `convert_to_tensor=False` and `convert_to_numpy=False`, then you skip all the extra ops after inference and are returned a List of Tensors to post-process as you please.

**Comment by nleroy917** [comment]
So, maybe they should both just default to `False`? So it's not opinionated and more efficient?

**Comment by mebrunet** [comment]
That makes sense to me. 

Also, I think having the option to pass in a preallocated Tensor of size `num_sentences` by `dim_embedding` that is progressively filled (batch by batch) would be better than relying on a final call to `stack`. But that's more work than just switching the defaults.

**Comment by nleroy917** [comment]
@nreimers are we correct here?

**Comment by Harshavardhan4123** [comment]
Is the issue resolved? Same problem for me even for a corpus size of 15
